neet/sensitivity.py

```diff
@@ -35,33 +39,178 @@ def sensitivity(net, state):
+                          when states argument is not None.
+    """
+    N = net.size
+    Q = np.zeros((N,N))
+
+    if (states is not None) or (weights is not None) or (not hasattr(net,'table')):
+        # explicitly calculate difference matrix for each state
+    
+        # optionally pre-calculate transitions
+        if calc_trans:
+            trans = list(transitions(net))
+        else:
+            trans = None
+    
+        # currently changes state generators to lists.
+        # is there a way to avoid this?
+        if states is None:
+            states = list(net.state_space())
+        else:
+            states = list(states)
+
+        if weights is None:
+            weights = np.ones(len(states))
+        else:
+            weights = list(weights)
+
+        if np.shape(weights) != (len(states),):
+            raise(ValueError("Weights must be a 1D array with length same as states"))
+
+        norm = np.sum(weights)
+        for i,state in enumerate(states):
+            Q += weights[i] * difference_matrix(net, state, trans) / norm
+
+    else: # make use of sparse connectivity to be more efficient
+        state0 = np.zeros(N,dtype=int)
+
+        for i in range(N):
+            nodesInfluencingI = _connections(net,i)
+            for jindex,j in enumerate(nodesInfluencingI):
+            
+                # for each state of other nodes, does j matter?
+                otherNodes = list(copy.copy(nodesInfluencingI))
+                otherNodes.pop(jindex)
+                otherNodeStates = _states_limited(otherNodes,state0)
+                for state in otherNodeStates:
+                    # (might be able to do faster by calculating transitions once
+                    #  for each i)
+                    # (also we only need the update for node i)
+                    # start with two states, one with j on and one with j off
+                    jOff = copy.copy(state)
+                    jOff[j] = 0
+                    jOffNext = net.update(jOff)[i]
+                    jOn = copy.copy(state)
+                    jOn[j] = 1
+                    jOnNext = net.update(jOn)[i]
+                    # are the results different?
+                    Q[i,j] += (jOffNext + jOnNext)%2
+                Q[i,j] /= float(len(otherNodeStates))
+            
+    return Q
+
+def lambdaQ(net,**kwargs):
+    """
+    Calculate sensitivity eigenvalue, the largest eigenvalue of the sensitivity
+    matrix average_difference_matrix.
+    
+    This is analogous to the eigenvalue calculated in
+    
+        Pomerance, A, E Ott, M Girvan, and W Losert. ``The Effect of Network
+        Topology on the Stability of Discrete State Models of Genetic Control.''
+        Proc. Natl. Acad. Sci. USA 106, no. 20 (2009): 8209-14.
+        http://www.pnas.org/content/106/20/8209.full.
+    """
+    Q = average_difference_matrix(net,**kwargs)
+    return max(np.sort(abs(linalg.eigvals(Q))))
+
+def _fast_encode(state):
```

Comment:
Why did you implement this? Is `StateSpace.encode` too slow? If so, maybe `StateSpace._unsafe_encode` will do?

Reply:
Ah, you're right, most of the speedup can be gotten by using _unsafe_encode.  But _fast_encode is still about 20% faster than _unsafe_encode on the cases I tried, so I think it's worth keeping in there.  (Either way, in networks around size 15, this encoding step is responsible for about a third of the total run time for calculating the average sensitivity...)